tests/scenario/test_mimir_cluster_interface.py

```diff
@@ -0,0 +1,108 @@
+from itertools import chain
+
+import ops
+import pytest
+from loki_cluster import (
+    LokiClusterProvider,
+    LokiClusterRequirerAppData,
+    LokiClusterRequirerUnitData,
+    LokiRole,
+)
+from ops import Framework
+from scenario import Context, Relation, State
+
+
+class MyCharm(ops.CharmBase):
+    META = {
+        "name": "lukasz",
+        "requires": {"loki-cluster-require": {"interface": "loki_cluster"}},
+        "provides": {"loki-cluster-provide": {"interface": "loki_cluster"}},
+    }
+
+    def __init__(self, framework: Framework):
+        super().__init__(framework)
+        self.provider = LokiClusterProvider(self, endpoint="loki-cluster-provide")
+
+
+@pytest.mark.parametrize(
+    "workers_roles, expected",
+    (
+        (
+            (({LokiRole.read}, 1), ({LokiRole.read}, 1)),
+            ({LokiRole.read: 2}),
+        ),
+        (
+            (({LokiRole.write}, 1), ({LokiRole.read}, 1)),
+            ({LokiRole.read: 1, LokiRole.write: 1}),
+        ),
+        ((({LokiRole.backend}, 2), ({LokiRole.backend}, 1)), ({LokiRole.backend: 3})),
+        (
+            (
+                ({LokiRole.read}, 2),
+                ({LokiRole.read}, 2),
+                ({LokiRole.read, LokiRole.backend}, 1),
+            ),
+            ({LokiRole.read: 5, LokiRole.backend: 1}),
+        ),
+    ),
+)
+def test_role_collection(workers_roles, expected):
```

Comment:
Seems like the functionality tested here should be ops-independent.
I.e seems like we shouldn't need harness/scenario to test this.

Reply:
Also, these behaviors are common to the shared objects, and should be tested in cos-lib rather than here